### PR TITLE
perf(*) plugins handlers do not have to inherit from BasePlugin anymore

### PR DESCRIPTION
**Pull request by thibaultcha**
Plugins handlers have always been inheriting from the `BasePlugin`
class, which is itself inheriting from the `Object` class provided by
`kong.vendor.classic`.

This design predates even the very first public release of Kong in 2015,
and was influenced by the OOP background of its original authors.

Today, the case can be made against this design:

- At best, it makes instances with `debug` log levels extremely noisy.
- At worst, it has a non-negligible performance impact due to the
  constant metatable lookups of a plugin's handlers methods and
  because of the non-JITable `ngx.log` calls sitting on hot code paths.
- All in all, its only purpose was to provide a `_name` attribute which
  was only used by the `BasePlugin` parent class and redundant with the
  `handler.name` property injected by the plugins loader.

This commit removes the need for plugins handlers to inherit from the
`BasePlugin` class, but still ensures backwards-compatibility for
plugins doing so.
Fixture plugins in our test suites were not updated and still inherit
from the `BasePlugin` class as a means to maintain regression tests for
this backwards-compatible behavior.

The `BasePlugin` is *not* deprecated by this commit.

In some initial benchmarks (sample size 3), this change increased
req/s by up to 6% when using a single plugin (key-auth). The more
plugins are configured for a request, the higher the gains.

### CODE REVIEW

File: kong/init.lua [hunk 1]
```diff
@@ -182,15 +182,15 @@ local function build_new_plugins(version)
   for _, plugin in ipairs(loaded_plugins) do
     if new_plugins.combos[plugin.name] then
       for phase_name, phase in pairs(new_plugins.phases) do
-        if plugin.handler[phase_name] ~= BasePlugin[phase_name] then
+        if type(plugin.handler[phase_name]) == "function"
```
Comment (hishamhm) [inline kong/init.lua]: just to make sure I follow: why was this change necessary, does any plugin declare a non-function field with a name clashing with a phase handler?

Comment (thibaultcha) [reply kong/init.lua]: The check could have been `if plugin.handler[phase_name] ~= nil`, but this is safer (most likely, plugins handlers should be subject to schema validation, just like migrations are, for instance).

If a handler is not implemented, the `plugin.handler[phase_name] ~= BasePlugin[phase_name]` condition would have been valid, and the phase incorrectly enabled by our runloop.

Comment (hishamhm) [reply kong/init.lua]: ahhh got it
